### PR TITLE
[EMP-21164] fix: Seeking not working for streams with ads

### PR DESCRIPTION
**Pull request by robert-pelka**
Refactored ServerSideAdService in order to fix a bug with seeking not possible for streams with server side ads.

### CODE REVIEW

File: Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift [hunk 1]
```diff
@@ -130,678 +106,597 @@ public class ServerSideAdService: AdService {
...
 
-
 // MARK: Playback
 extension ServerSideAdService {
+    private func startObservingPlayer(originalScrubPosition: Int64, targetScrubPosition: Int64) {
+        var originalPosition = originalScrubPosition
+        var targetPosition = targetScrubPosition
+        
+        self.tech.addPeriodicTimeObserverToPlayer { [weak self] time in
+            guard let `self` = self else {
```
Comment (jkochaniec) [inline Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift:182]: `guard let self else {` - should be enough here

File: Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift [hunk 2]
```diff
@@ -130,678 +106,597 @@ public class ServerSideAdService: AdService {
...
         
-        // Add periodic time observer for the player
-        self.tech.addPeriodicTimeObserverToPlayer { [weak self] time in
-            
-            guard let `self` = self else { return }
+        guard let timeInMil = time.milliseconds?.rounded() else {
+            return
+        }
+        
+        for (index, content) in self.allTimelineContent.enumerated() {
```
Comment (jkochaniec) [inline Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift:252]: calling `self` seems to be unneeded here

File: Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift [hunk 3]
```diff
@@ -130,678 +106,597 @@ public class ServerSideAdService: AdService {
...
-                        rangeEnd = 0
-                    }
-                } else {
-                    // No `AdBreak` was found, keep playing the content
-                    self.userInitiatedSeek = true
-                    rangeStart = 0
-                    rangeEnd = 0
-                }
-                
+            if !self.alreadyPlayedAds.contains(content) {
```
Comment (jkochaniec) [inline Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift:262]: no need to use `self` here

File: Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift [hunk 4]
```diff
@@ -130,678 +106,597 @@ public class ServerSideAdService: AdService {
...
-                                
-                                let clip = self.clips[adClipIndex]
-
-                                if (timeInMil / 10 * 10) == (clipFirstQuartile / 10 * 10)  {
+                handleAlreadyWatchedAd(index, content)
+            }
+        }
+    }
+    
+    private func handleNotWatchedYetAd(_ content: TimelineContent, _ start: Int64, _ end: Int64, _ timeInMil: Int64) {
```
Comment (jkochaniec) [inline Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift:270]: I would rename `timeInMil` to something more descriptive `timeInMilliseconds`

File: Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift [hunk 5]
```diff
@@ -130,678 +106,597 @@ public class ServerSideAdService: AdService {
...
     
-                                    // Send firstQuartile tracking events
-                                    if let firstQuartileUrls = clip.trackingEvents?.firstQuartile {
-                                        if !self.alreadySentadTrackingUrls.contains(firstQuartileUrls) {
-                                            self.context.trackAds(adTrackingUrls: firstQuartileUrls)
-                                            self.alreadySentadTrackingUrls.append( firstQuartileUrls )
-                                        }
-                                    }
+    private func sendTrackingEvents(_ point: [String]?) {
+        guard let point,
```
Comment (jkochaniec) [inline Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift:300]: unneeded `self` usage inside function

File: Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift [hunk 6]
```diff
@@ -130,678 +106,597 @@ public class ServerSideAdService: AdService {
...
-                                }
-                            }
-                        }
-                        else {
-                            return
-                        }
-                    }
-                }
+    private func handleAdEnd(_ clip: AdClips, _ content: TimelineContent) {
+        // Send EMP analytics
```
Comment (jkochaniec) [inline Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift:311]: unneeded self usage inside function

**Review by jkochaniec — CHANGES_REQUESTED** [review]

File: Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift [hunk 7]
```diff
@@ -291,140 +294,141 @@ extension ServerSideAdService {
...
     
     private func handleAdStartWithDelay(
         _ content: TimelineContent,
         _ start: Int64,
         _ end: Int64,
-        _ timeInMil: Int64,
+        _ timeInMiliseconds: Int64,
         _ adClipIndex: Array<TimelineContent>.Index
     ) {
         // Some content may not have the `timeInMil` 0 when start. It seems like `PeriodicTimeObserverToPlayer` may have a slight delay and then timeInMil may be higher than the `start`. [ Add a second condition to check if timeInMil is larger than start :=> Ad has already started] . But this will only run once when the ad has started.
```
Comment (jkochaniec) [inline Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift:349]: `timeInMil` - > `timeInMiliseconds` in comment

File: Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift [hunk 8]
```diff
@@ -291,140 +294,141 @@ extension ServerSideAdService {
...
-        guard timeInMil.rounded() + 10 == start.rounded() + 10 || timeInMil.rounded() + 10 > start.rounded() + 10,
-              !self.alreadyStartedAds.contains(content)
+        guard timeInMiliseconds.rounded() + 10 == start.rounded() + 10 ||
```
Comment (jkochaniec) [inline Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift:350]: Can we name 10 somehow?

**Review by jkochaniec — CHANGES_REQUESTED** [review]

File: Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift (hunk 5 above)
Comment (robert-pelka) [reply Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift:298]: removed

File: Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift (hunk 6 above)
Comment (robert-pelka) [reply Sources/iOSClientExposurePlayback/Services/AdService/ServerSideAdService.swift:309]: removed

**Review by jkochaniec — APPROVED** [review]

**Pull request merged**